atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraBackedPueTableTest.java

```diff
@@ -0,0 +1,93 @@
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.atlasdb.containers.CassandraResource;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.pue.ConsensusForgettingStore;
+import com.palantir.atlasdb.pue.KvsConsensusForgettingStore;
+import com.palantir.atlasdb.pue.PutUnlessExistsTable;
+import com.palantir.atlasdb.pue.ResilientCommitTimestampPutUnlessExistsTable;
+import com.palantir.atlasdb.table.description.TableMetadata;
+import com.palantir.atlasdb.transaction.encoding.TwoPhaseEncodingStrategy;
+import com.palantir.atlasdb.transaction.impl.TransactionConstants;
+import com.palantir.common.concurrent.PTExecutors;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+public class CassandraBackedPueTableTest {
+    private final KeyValueService kvs = CASSANDRA.getDefaultKvs();
+    private final ConsensusForgettingStore store =
+            new KvsConsensusForgettingStore(kvs, TransactionConstants.TRANSACTIONS2_TABLE);
+    private final PutUnlessExistsTable<Long, Long> pueTable =
+            new ResilientCommitTimestampPutUnlessExistsTable(store, TwoPhaseEncodingStrategy.INSTANCE);
+
+    @ClassRule
+    public static final CassandraResource CASSANDRA = new CassandraResource();
+
+    @Before
+    public void setup() {
+        kvs.createTable(
+                TransactionConstants.TRANSACTIONS2_TABLE,
+                TableMetadata.allDefault().persistToBytes());
+    }
+
+    @Test
+    public void getDoesNotLeakExceptionsInRaceConditions() throws ExecutionException, InterruptedException {
+        ExecutorService writeExecutor = PTExecutors.newFixedThreadPool(1);
+        ExecutorService readExecutors = PTExecutors.newFixedThreadPool(10);
+
+        List<Long> timestamps = LongStream.range(0, 500).mapToObj(x -> x * 100).collect(Collectors.toList());
+        Iterable<List<Long>> partitionedStartTimestamps = Lists.partition(timestamps, 20);
+        for (List<Long> singlePartition : partitionedStartTimestamps) {
+            singlePartition.forEach(
+                    timestamp -> writeExecutor.execute(() -> pueTable.putUnlessExists(timestamp, timestamp)));
+
+            List<Future<ListenableFuture<Map<Long, Long>>>> reads = new ArrayList<>();
+            for (int i = 0; i < singlePartition.size(); i++) {
+                reads.add(readExecutors.submit(() -> pueTable.get(singlePartition)));
```

Comment:
eh, this seems a bit sus (there's a part of me that says we should be able to achieve this by passing in the executor into the Futures.transform()), but fine given time pressure.